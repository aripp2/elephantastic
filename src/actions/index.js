export const setRandom = pup =>({
  type: 'SET_RANDOM',
  pup
})

export const throwError = error => ({
  type: 'THROW_ERROR',
  error
})

export const setFavs = favs => ({
  type: 'SET_FAVS',
  favs
})