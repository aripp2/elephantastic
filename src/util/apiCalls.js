const xKey = 'fdae1377-8b39-4c2b-a87f-38d9ef268562';
const getOptions = {
    method: 'GET',
    headers: {
      'x-api-key': xKey
    }
  };

export const getRandom = async() => {
  const url = 'https://api.thedogapi.com/v1/images/search';
  const response = await fetch(url, getOptions)
  if(!response.ok) {
    throw Error('There was an issue getting a pup, try again later.')
  }
  const random = await response.json();
  return random[0]

}

export const postVote = async (imgId, value) => {
  const url = 'https://api.thedogapi.com/v1/votes';
  const options = {
    method: 'POST',
    body: JSON.stringify({
      image_id: imgId,
      value: value
    }),
    headers: {
      'Content-type': 'application/json',
      'x-api-key': xKey
    }
  }

  const response = await fetch(url, options);
  if(!response.ok) {
    throw Error('Unable to submit your vote.')  
  }
  const votes = await response.json()
  console.log('postVote in apiCalls', votes)
}

export const getBreeds = async () => {
  const url = 'https://api.thedogapi.com/v1/breeds';
  const response = await fetch(url, getOptions);
  if(!response.ok) {
    console.log(response)
  }
  const breeds = await response.json();
  // console.log('getBreeds in apiCalls', breeds)
  return breeds;
}

export const getFavorites = async() => {
  const url = 'https://api.thedogapi.com/v1/favourites';
  const response = await fetch(url, getOptions)
  if(!response.ok) {
    console.log(response)
  }
  const favs = await response.json();
  console.log('favs', favs)
  return favs;
}

export const addFavorite = async(imgId) => {
  const url = 'https://api.thedogapi.com/v1/favourites';
  const options = {
    method: 'POST',
    body: JSON.stringify({
      image_id: imgId
    }),
    headers: {
      'Content-type': 'application/json',
      'x-api-key': xKey
    }
  } 

  const response = await fetch(url, options)
  if(!response.ok) {
    throw Error('Unable to add favorite.')
  }
}

export const deleteFavorite = async(favId) => {
  const url = `https://api.thedogapi.com/v1/favourites/${favId}`
  const options = {
    method: 'DELETE',
    headers: {
      'Content-type': 'application/json',
      'x-api-key': xKey
    }
  }
  const response = await fetch(url, options)
  if(!response.ok) {
    throw Error('Unable to remove this favorite.')
  }
}

export const getBreedImages = async(breedId) => {
  const url = `https://api.TheDogAPI.com/images/search?breed_id=${breedId}`
  const response = await fetch(url, getOptions)
  if(!response.ok) {
    console.log('selected breed response', response)
  }
  const selected = await response.json()
  console.log('selected in api', selected)
  return selected
}





