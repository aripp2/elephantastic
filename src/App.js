import React, { Component } from 'react';
import { getRandom } from './util/apiCalls';
import Header from './Header/Header';
import NavForm from './NavForm/NavForm';
import MainContainer from './MainContainer/MainContainer';
import './App.scss';

export class App extends Component {
  constructor() {
    super();
    this.state = {
      random: {},
      error: ''
    }
  }

  async componentDidMount() {
    try {
      const random = await getRandom();
      this.setState({ random })
    } catch({ message }) {
      this.setState({ message})
    }
  }

  render() {


    return (
      <div className="App">
        <Header />
        <main>
          <NavForm />
          <MainContainer random={this.state.random}/>
        </main>
      </div>
    );
  }
}

export default App;
