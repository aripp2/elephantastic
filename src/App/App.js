import React, { Component } from 'react';
import { Route } from 'react-router-dom';
import { connect } from 'react-redux';
import { getRandom, getFavorites, addFavorite, deleteFavorite, getBreeds } from '../util/apiCalls';
import { setRandom, throwError, updateLoading, setFavs, setBreeds } from '../actions';
import NavHeader from '../NavHeader/NavHeader';
import SearchForm from '../SearchForm/SearchForm';
import SearchContainer from '../SearchContainer/SearchContainer';
import FavoritesContainer from '../FavoritesContainer/FavoritesContainer';
import Vote from '../Vote/Vote';
import './App.scss';

export class App extends Component {
  constructor() {
    super();
    this.state = {
    }
  }

  updateRandom = async() => {
    const { setRandom, throwError, updateLoading } = this.props;
    try {
      const random = await getRandom();
      setRandom(random)
      updateLoading(false)
    } catch({ message }) {
      updateLoading(false)
      throwError(message)
    }
  }

  async componentDidMount() {
    const { setFavs, throwError, updateLoading, setBreeds } = this.props
    try {
      const favs = await getFavorites()
      setFavs(favs)
      await this.updateRandom()
      const breeds = await getBreeds()
      setBreeds(breeds)
      updateLoading(false)
    } catch ({ message }) {
      updateLoading(false)
      throwError(message)
    }
  }

  updateFavs = async(id, status, favId) => {
    const { setFavs, throwError, updateLoading } = this.props;
    updateLoading(true)
    try {
      if (status) {
        await deleteFavorite(favId)
      } else {
        await addFavorite(id)
      }
        const favs = await getFavorites()
        setFavs(favs)
        updateLoading(false)
    } catch ({ message }){
      updateLoading(false)
      throwError(message)
    }
  }

  render() {
    const { errorMsg, isLoading, randomPup, favorites } = this.props;
    console.log(randomPup)
    return (
      <div className="App">
        <NavHeader />
        <main>
          {errorMsg && <h2>{errorMsg}</h2>}
          {isLoading && <h2>Loading...</h2>}
          {!isLoading && <Route exact path='/' render={() => <Vote updateFavs={this.updateFavs} updateRandom={this.updateRandom}/>}/>}
          <Route path='/search' render={() => 
              <SearchForm />}/>
          <Route path='/favorites' render={() => 
            <FavoritesContainer updateFavs={this.updateFavs}/>} />

        </main>
      </div>
    );
  }
}

export const mapStateToProps = ({ errorMsg, isLoading, randomPup, favorites }) => ({
  errorMsg,
  isLoading,
  randomPup,
  favorites
})

export const mapDispatchToProps = dispatch => ({
  setRandom: pup => dispatch(setRandom(pup)),
  throwError: error => dispatch(throwError(error)),
  setFavs: favs => dispatch(setFavs(favs)),
  updateLoading: bool => dispatch(updateLoading(bool)),
  setBreeds: breeds => dispatch(setBreeds(breeds))
})

export default connect(mapStateToProps, mapDispatchToProps)(App);
