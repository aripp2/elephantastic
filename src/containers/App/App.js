import React, { Component } from 'react';
import { Route } from 'react-router-dom';
import { connect } from 'react-redux';
import { getRandom, getFavorites, addFavorite, deleteFavorite, getBreeds } from '../../util/apiCalls';
import { setRandom, throwError, updateLoading, setFavs, setBreeds } from '../../actions';
import NavHeader from '../../components/NavHeader/NavHeader';
import SearchForm from '../SearchForm/SearchForm';
import FavoritesContainer from '../FavoritesContainer/FavoritesContainer';
import Vote from '../Vote/Vote';
import './App.scss';

export class App extends Component {
  constructor() {
    super();
    this.state = {
    }
  }

  updateRandom = async() => {
    const { setRandom, throwError, updateLoading } = this.props;
    updateLoading(true)
    try {
      const random = await getRandom();
      setRandom(random)
      updateLoading(false)
    } catch({ message }) {
      updateLoading(false)
      throwError(message)
    }
  }

  async componentDidMount() {
    const { setFavs, throwError, updateLoading, setBreeds } = this.props
    try {
      const favs = await getFavorites()
      setFavs(favs)
      await this.updateRandom()
      const breeds = await getBreeds()
      setBreeds(breeds)
      updateLoading(false)
    } catch ({ message }) {
      updateLoading(false)
      throwError(message)
    }
  }

  updateFavs = async(id, status, favId) => {
    const { setFavs, throwError, updateLoading } = this.props;
    updateLoading(true)
    try {
      if (status) {
        await deleteFavorite(favId)
      } else {
        await addFavorite(id)
      }
        const favs = await getFavorites()
        setFavs(favs)
        updateLoading(false)
    } catch ({ message }){
      updateLoading(false)
      throwError(message)
    }
  }

  render() {
    const { errorMsg, isLoading } = this.props;
    return (
      <div className="App">
        <NavHeader />
        <main>
          {errorMsg && <h2>{errorMsg}</h2>}
          {isLoading && <img src='https://cdn.dribbble.com/users/5976/screenshots/3930514/happy_dog_puppy_tongue_logo_design_symbol_by_alex_tass.gif' alt='loading gif'/>}
          {!isLoading && <Route exact path='/' render={() => <Vote updateFavs={this.updateFavs} updateRandom={this.updateRandom}/>}/>}
          <Route path='/search' render={() => 
              <SearchForm />}/>
          <Route path='/favorites' render={() => 
            <FavoritesContainer updateFavs={this.updateFavs}/>} />

        </main>
      </div>
    );
  }
}

export const mapStateToProps = ({ errorMsg, isLoading }) => ({
  errorMsg,
  isLoading
})

export const mapDispatchToProps = dispatch => ({
  setRandom: pup => dispatch(setRandom(pup)),
  throwError: error => dispatch(throwError(error)),
  setFavs: favs => dispatch(setFavs(favs)),
  updateLoading: bool => dispatch(updateLoading(bool)),
  setBreeds: breeds => dispatch(setBreeds(breeds))
})

export default connect(mapStateToProps, mapDispatchToProps)(App);
