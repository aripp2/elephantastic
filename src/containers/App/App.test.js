import React from 'react';
import { shallow } from 'enzyme';
import { App, mapStateToProps, mapDispatchToProps } from './App';
import { setRandom, throwError, updateLoading, setFavs, setBreeds } from '../../actions';
import { getRandom, getFavorites, addFavorite, deleteFavorite, getBreeds } from '../../util/apiCalls';


jest.mock('../../util/apiCalls.js');

describe('App', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />)
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

});

describe('mapStateToProps', () => {

  it('should return an object of errorMsg and isLoading', () => {
    const mockState = {
      favorites: [],
      randomPup: {},
      errorMsg: '',
      isLoading: false,
      breeds: []
    };
    const expected = {
      errorMsg: '',
      isLoading: false
    }
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {
  const mockDispatch = jest.fn();

  it('calls dispatch with setRandom action when updateRandom is called', () => {
    const mockRandomPup = {
      breeds: [],
      height: 720,
      id: "anIClTXyD",
      url: "https://cdn2.thedogapi.com/images/anIClTXyD.jpg",
      width: 350
    }
    const actionToDispatch = setRandom(mockRandomPup);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.updateRandom()
    expect(mockDispatch).toHaveBeenCalledWith(mockRandomPup)
  });

});


// describe('mapDispatchToProps', () => {
//     it('calls dispatch with an addTodo action when handleSubmit is called', () => {
//       // Setup
//       const mockDispatch = jest.fn();
//       const actionToDispatch = addTodo('Learn Redux!', 1);

//       // Execution
//       const mappedProps = mapDispatchToProps(mockDispatch);
//       mappedProps.handleSubmit('Learn Redux!', 1);

//       // Expectaion
//       expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
//     });
//   });
