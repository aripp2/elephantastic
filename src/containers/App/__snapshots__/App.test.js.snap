// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="App"
>
  <NavHeader />
  <main>
    <Route
      exact={true}
      path="/"
      render={[Function]}
    />
    <Route
      path="/search"
      render={[Function]}
    />
    <Route
      path="/favorites"
      render={[Function]}
    />
  </main>
</div>
`;
